### PR TITLE
Implement tracing

### PR DESCRIPTION
**Pull request by angelala3252**
This PR addresses [issue #170 ](https://github.com/MLDSAI/OpenAdapt/issues/170) by adding the @logging decorator shown in [this article](https://blog.koehntopp.info/2023/03/14/tracing-python.html). I changed the print statements to logging as it got a little messy with print statements alternating with the logging, and I also got rid of the indentation as it didn't seem necessary when using logging. I didn't add the @logging decorator to the functions that got triggered a lot like the event handlers. 

Next steps:
- the `record` function still hangs sometimes near the end without any additional info from the added tracing, so I will have to find another way to give more information about what is causing the issue
- even without the tracing, all the logging from different threads often gets mixed together into one sentence (example shown in the image below). If we want to make the logging more readable, we will have to make sure that only one thread logs at one time.

image

Instructions to run:
``` 
git remote add angelala3252 https://github.com/angelala3252/OpenAdapt.git
git fetch angelala3252
git checkout feat/tracing
python -m openadapt.record "test" 
# tracing will show up in the terminal, "Enter" for when you enter a function, and "Leave" for when a function returns
```

### CODE REVIEW

**Comment by angelala3252** [comment]
Update:
The log messages now no longer get mixed up as they are now multiprocess safe. The recording hanging issue was due to a separate error I was getting, so now the recording does not hang and every major function gets traced. I haven't gone any further with the tracing itself, as it still just shows entering and exiting functions, but if that is necessary I can explore that more.

**Comment by abrichr** [comment]
@angelala3252 can you please fix merge conflicts?

**Comment by angelala3252** [comment]
@abrichr resolved!

File: openadapt/record.py [hunk 1]
```diff
@@ -37,6 +39,43 @@
...
+
+
+def args_to_str(*args):
+    return ", ".join(map(str, args))
+
+
+def kwargs_to_str(**kwargs):
+    ret = ""
+    for k, v in kwargs.items():
+        ret += f"{k}={v},"
```
Comment (abrichr) [inline openadapt/record.py:52]: What do you think about:
```
return ",".join([f"{k}={v}" for k, v in kwargs.items()])
```
🤓

File: openadapt/record.py [hunk 2]
```diff
@@ -37,6 +39,43 @@
...
+
+    return ret[:-1]
+
+
+def logging(logger):
```
Comment (abrichr) [inline openadapt/record.py:57]: Maybe we should name this differently, e.g. `trace(logger)`, what do you think?

File: openadapt/utils.py [hunk 3]
```diff
@@ -26,9 +27,16 @@ def configure_logging(logger, log_level):
     log_level_override = os.getenv("LOG_LEVEL")
     log_level = log_level_override or log_level
     logger.remove()
+    logger_format = ("<green>{time:YYYY-MM-DD HH:mm:ss.SSS}</green> | "
+                     "<level>{level: <8}</level> | "
+                     "<cyan>{name}</cyan>:<cyan>{function}</cyan>:<cyan>{line}</cyan> "
+                     "- <level>{message}</level>")
```
Comment (abrichr) [inline openadapt/utils.py:33]: Can you please format this as follows:
```
    logger_format = (
        "<green>{time:YYYY-MM-DD HH:mm:ss.SSS}</green> | "
         "<level>{level: <8}</level> | "
         "<cyan>{name}</cyan>:<cyan>{function}</cyan>:<cyan>{line}</cyan> "
         "- <level>{message}</level>"
)
```
Also can you please clarify the origins of this format?

File: openadapt/record.py (hunk 2 above)
Comment (angelala3252) [reply openadapt/record.py]: renamed in [db24c49](https://github.com/MLDSAI/OpenAdapt/pull/229/commits/db24c49d6d83631b825b6706c3efa316f9ac5030)

**Pull request merged**